microsoft/testsuites/dpdk/dpdksuite.py

```diff
@@ -0,0 +1,242 @@
+        mount.mount("nodev", "/mnt/huge-1G", type="hugetlbfs", options="'pagesize=1G'")
+
+    def _hugepages_enable(self, node: Node, log: Logger) -> None:
+        self._execute_expect_zero(
+            node,
+            (
+                "sh -c 'echo 4096 > "
+                "/sys/devices/system/node/node0/hugepages/"
+                "hugepages-2048kB/nr_hugepages'"
+            ),
+        )
+        self._execute_expect_zero(
+            node,
+            (
+                "sh -c 'echo 1 > "
+                "/sys/devices/system/node/node0/hugepages/"
+                "hugepages-1048576kB/nr_hugepages'"
+            ),
+        )
+        node.execute("grep -i huge /proc/meminfo && ls /mnt/", shell=True)
+
+
+class NicInfo:
+
+    # Class for info about an single upper/lower nic pair.
+    # devices using SRIOV on azure typically have an upper synthetic device
+    # paired with a lower SRIOV Virtual Function (VF) device that
+    # enables the passthrough to the physical NIC.
+
+    def __init__(self, upper: str, lower: str = "", pci_slot: str = "") -> None:
+        self._has_secondary = lower != "" or pci_slot == ""
+        self._upper = upper
+        self._lower = lower
+        self._pci_slot = pci_slot
+
+    def has_secondary(self) -> bool:
+        return self._has_secondary
+
+    def __str__(self) -> str:
+        return (
+            "NicInfo:\n"
+            f"upper: {self._upper}\n"
+            f"lower: {self._lower}\n"
+            f"pci_slot: {self._pci_slot}\n"
+        )
+
+    def testpmd_include(self, vdev_type: str) -> str:
+        assert_that(self._has_secondary).is_true().described_as(
+            (
+                f"This interface {self._upper} does not have a lower interface "
+                "and pci slot associated with it. Aborting."
+            )
+        )
+        return f'--vdev="{vdev_type},iface={self._upper}" --allow "{self._pci_slot}"'
+
+
+class NodeNicInfo:
+
+    # Class for all of the nics on a node. Contains multiple NodeNic classes.
+    # Init identifies upper/lower paired devices and the pci slot info for the lower.
+
+    _pci_slot_regex = r"([0-9a-fA-F]{1-4}:)+[0-9a-fA-F]{1-4}"
+    _nics: Dict[str, NicInfo] = dict()
+
+    def _get_nic_names(self, node: Node) -> List[str]:
+        # identify all of the nics on the device, excluding tunnels and loopbacks etc.
+        result = node.execute(
+            " ls /sys/class/net/ | grep -Ev $(ls /sys/devices/virtual/net)",
+            shell=True,
+            sudo=True,
+        )
+        nic_names = result.stdout.split("\r\n")
+        for item in nic_names:
+            assert_that(item).is_not_equal_to("").described_as(
+                "nic name could not be found"
+            )
+        return nic_names
+
+    def _get_nic_device(self, node: Node, nic_name: str) -> str:
+        slot_info_result = node.execute(f"readlink /sys/class/net/{nic_name}/device")
+        assert_that(slot_info_result.exit_code).is_zero()
```

Comment:
call `slot_info_result.assert_exit_code`

Reply:
Will change, I feel like node.execute should just have a keyword arg for expected error code and optionally turn off assertions if you don't care. It's much rarer to not expect a 0 error code or not care about the result so this could get rid of a lot of extra code if we just asserted on it by default and let the user turn the assert off if they want or change the error code if they want.

Reply:
Yes, we can have that. I will add such an functionality. And it's off by default.